### PR TITLE
fix(motions): improve compat for jsdom & jest

### LINKED ISSUE #31593: [Bug]: Dialog remains in DOM after close when using Jest
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: Linux 5.15 Ubuntu 22.04.4 LTS 22.04.4 LTS (Jammy Jellyfish)
    CPU: (16) x64 Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
    Memory: 28.73 GB / 31.32 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Browsers:
    Chrome: 123.0.6312.105
```

### Are you reporting Accessibility issue?

no

### Reproduction

Add Dialog close validation to `packages/react-components/react-dialog/src/components/Dialog/Dialog.test.tsx`

```diff
    const result = render(
      <Dialog>
        <DialogTrigger disableButtonEnhancement>
          <button data-testid="trigger">Open dialog</button>
        </DialogTrigger>
        <CustomDialogSurface>
          <div data-testid="surface-content">content in surface</div>
+          <DialogTrigger>
+            <button data-testid="trigger-close">Close dialog</button>
+          </DialogTrigger>
        </CustomDialogSurface>
      </Dialog>,
    );

    result.getByTestId('trigger').click();

    expect(result.getByTestId('surface-content')).toBeInTheDocument();
    expect(result.getByTestId('surface-content')).toBeVisible();

+    result.getByTestId('trigger-close').click();
+    expect(result.queryByTestId('surface-content')).toBeNull();
```

Then run the tests.

## Actual Behavior
The test passed. The element with `data-testid="surface-content"` does not exist. `react-components@9.52.0` passes this test.

## Expected Behavior
The test failed. The element with `data-testid="surface-content"` exists. `react-components@9.53.0` fails this test.

### Logs

_No response_

### Requested priority

Blocking

### Products/sites affected

ODSP

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by layershifter**
## New Behavior

Improves compat with Jest & JSDOM, so tests would pass as expected.

## Related Issue(s)

Fixes #31593.

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-mount_base.html">85</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-mount.html">96</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">37</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">30</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Avatar-mount_base.html">646</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Avatar-mount.html">648</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Button-mount_base.html">303</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Button-mount.html">298</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Field-mount_base.html">1134</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Field-mount.html">1117</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProvider-mount_base.html">709</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProvider-mount.html">708</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-mount_base.html">85</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-mount.html">96</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">37</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">30</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">82</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">77</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/MakeStyles-mount_base.html">860</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/MakeStyles-mount.html">869</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Persona-mount_base.html">1749</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/Persona-mount.html">1724</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/SpinButton-mount_base.html">1407</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/SpinButton-mount.html">1394</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/SwatchPicker-mount_base.html">1520</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31602/perf-test-react-components/SwatchPicker-mount.html">1597</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: change/@fluentui-react-motion-b068209f-dd7d-43e0-a77a-f6ab0606659d.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-motion-b068209f-dd7d-43e0-a77a-f6ab0606659d.json:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31602&commitId=ea80d1be1040c40e62a41c6379ccea38298bfb3a&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask - RTL.normal.chromium.png|2| Changed|
|Avatar Converged.size+image+badge.normal.chromium.png|0| Removed|

</details>


  <details><summary><b>Card Converged - Interactive</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Interactive.appearance interactive - Filled - Dark Mode.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled - Dark Mode.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Filled - Dark Mode.normal.chromium.png|0| Added|

</details>


  <details><summary><b>Card Converged - Selectable</b> 4 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Selectable.appearance selectable - Filled Alternative - RTL.click.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Filled Alternative - RTL.hover.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Filled Alternative - RTL.normal.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Filled Alternative - RTL.selected.chromium.png|0| Removed|

</details>


  <details><summary><b>SwatchPicker Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|SwatchPicker Converged.Spacing.default.chromium.png|0| Added|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-motion/src/utils/animateAtoms.ts [hunk 2]
```diff
@@ -38,6 +38,18 @@ export function animateAtoms(
       });
     },
     set onfinish(callback: () => void) {
+      // Heads up!
+      // Jest uses jsdom as the default environment, which doesn't support the Web Animations API. This no-op is
+      // necessary to avoid errors in tests.
+      //
+      // See https://github.com/microsoft/fluentui/issues/31593
```
Comment (ling1726) [inline packages/react-components/react-motion/src/utils/animateAtoms.ts:45]: is there an issue in JSDOM for this too?

**Review by ling1726 — APPROVED** [review]

File: packages/react-components/react-motion/src/utils/animateAtoms.ts (hunk 2 above)
Comment (layershifter) [reply packages/react-components/react-motion/src/utils/animateAtoms.ts:45]: yes

Comment (layershifter) [reply packages/react-components/react-motion/src/utils/animateAtoms.ts:45]:
```suggestion
      // See https://github.com/microsoft/fluentui/issues/31593
      // See https://github.com/jsdom/jsdom/issues/3429
```

**Pull request merged**